graphene_sqlalchemy/converter.py

```diff
@@ -26,29 +33,46 @@ def is_column_nullable(column):
     return bool(getattr(column, "nullable", True))
 
 
-def convert_sqlalchemy_relationship(relationship_prop, registry, connection_field_factory, resolver, **field_kwargs):
-    direction = relationship_prop.direction
-    model = relationship_prop.mapper.entity
-
+def convert_sqlalchemy_relationship(relationship_prop, obj_type, connection_field_factory, batching,
+                                    attr_name, orm_field_name, **field_kwargs):
+    """
+    :param sqlalchemy.RelationshipProperty relationship_prop:
+    :param Registry registry:
+    :type function|None connection_field_factory:
+    :type bool batching:
+    :param SQLAlchemyObjectType obj_type:
+    :param str orm_field_name:
+    :rtype: Dynamic
+    """
     def dynamic_type():
-        _type = registry.get_type_for_model(model)
+        direction = relationship_prop.direction
+        model = relationship_prop.mapper.entity
+        type_ = obj_type._meta.registry.get_type_for_model(model)
 
-        if not _type:
+        batching_ = batching if is_selectin_available else False
+        connection_field_factory_ = connection_field_factory
+
+        if not type_:
             return None
+
         if direction == interfaces.MANYTOONE or not relationship_prop.uselist:
-            return Field(
-                _type,
-                resolver=resolver,
-                **field_kwargs
-            )
-        elif direction in (interfaces.ONETOMANY, interfaces.MANYTOMANY):
-            if _type._meta.connection:
-                # TODO Add a way to override connection_field_factory
-                return connection_field_factory(relationship_prop, registry, **field_kwargs)
-            return Field(
-                List(_type),
-                **field_kwargs
-            )
+            resolver = get_custom_resolver(obj_type, orm_field_name)
+            if resolver is None:
```

Comment:
I don't find this syntax very readable. maybe something like that ?
```suggestion
            if resolver is None and batching_: 
```

Reply:
That would be a different behavior. To get the same behavior, I would need to do something like that:
```py
if resolver is None and batching_:
  resolver = get_batch_resolver(relationship_prop)
elif resolver is None and not batching_:
   resolver = get_attr_resolver(obj_type, relationship_prop.key)
```
I think I prefer the current version because it reads to me more like the intended logic:
> If no custom resolver was defined by the user, 
> then set it to batch resolver if batching is enabled, otherwise set it to the regular resolver

That said, I don't feel strongly about it so let me know what you prefer.

Reply:
Yeah, I understand your point, indeed
The main problem, finally, is the "complexity" due to multiple indentation, which could resolve by a more "splited" code.

Maybe not in this PR, but this block could be splited in smaller chunks.

Reply:
I see. I tried to extract some logic into sub-functions it but I'm not sure I like this because those sub-functions don't seem to make much sense on their own: 
```py
def convert_sqlalchemy_relationship(relationship_prop, obj_type, connection_field_factory, batching,
                                    attr_name, orm_field_name, **field_kwargs):
    def dynamic_type():
        direction = relationship_prop.direction
        model = relationship_prop.mapper.entity
        type_ = obj_type._meta.registry.get_type_for_model(model)

        batching_ = batching if is_selectin_available else False
        connection_field_factory_ = connection_field_factory

        if not type_:
            return None

        if direction == interfaces.MANYTOONE or not relationship_prop.uselist:
            resolver = _get_resolver_for_o2o_or_m2o_relationship(obj_type, orm_field_name, relationship_prop, batching_)
            return Field(type_, resolver=resolver, **field_kwargs)

        if direction in (interfaces.ONETOMANY, interfaces.MANYTOMANY):
            if not type_._meta.connection:
                return Field(List(type_), **field_kwargs)

            connection_field_factory_ = _get_connection_field_factory(connection_field_factory_, batching_)
            return connection_field_factory_(relationship_prop, obj_type._meta.registry, **field_kwargs)

    return Dynamic(dynamic_type)


def _get_resolver_for_o2o_or_m2o_relationship(obj_type, orm_field_name, relationship_prop, batching):
    resolver = get_custom_resolver(obj_type, orm_field_name)
    if resolver is None:
        return get_batch_resolver(relationship_prop) if batching else \
            get_attr_resolver(obj_type, relationship_prop.key)


def _get_connection_field_factory(connection_field_factory, batching):
    # TODO Allow override of connection_field_factory and resolver via ORMField
    if connection_field_factory is None:
        connection_field_factory = BatchSQLAlchemyConnectionField.from_relationship if batching else \
            default_connection_field_factory
    
    return connection_field_factory
```
You really need to understand the logic of `convert_sqlalchemy_relationship` to understand `_get_resolver_for_o2o_or_m2o_relationship` and `_get_connection_field_factory`. That said, it makes `convert_sqlalchemy_relationship` reading a bit easier in the sense that you can clearly see the 3 main `if` statements. 

What are you thoughts? I could go either way on this.

Reply:
Yeah, indeed these 2 methods are not really understandable on their own, but as it's spliting the code and allowing a easier reading, I think it could be a good "first step" for a big improvement : the next time someone will modify this code, it will be easier to go a step further

Reply:
Fair point. I ended up going a step further.